Display UPE fields in checkout with less than 50 cents in the cart

RadoslavGeorgiev opened the pull request:
Fixes #2700

### Description of the issue

In order to avoid money loss, Stripe has [minimum and maximum charge amounts](https://stripe.com/docs/currencies#minimum-and-maximum-charge-amounts).

When using Stripe Elements (not UPE) Stripe's API returned an error (see image below) when attempting to create and confirm the intent, which happens after the checkout form is submitted.

![classic-error](https://user-images.githubusercontent.com/5311119/132677646-dd251fcb-2586-4107-9c75-bbfa9d967ae1.png)

This was acceptable, because it allowed fields to be displayed on the checkout page **and a shipping method to be selected**, which might bring the order total above the minimum required amount.

As noted in https://github.com/Automattic/woocommerce-payments/issues/2700, UPE fields require a payment intent to be created in order to use the secret of that intent for displaying the fields. Whenever it happens that the **cart total** is lower than the minimum amount, payment fields would not be displayed at all.

### The purpose of this PR is to:

- Allow UPE fields to be displayed even when the cart total is less than 50 cents (or the amount for the store currency).
- Include more information in the error message both for UPE fields and for normal fields:

<img width="803" alt="image" src="https://user-images.githubusercontent.com/5311119/132678797-dff98a6a-43c3-498a-b67a-10b121f653d4.png">

This message indicates to customers what necessitates the minimum amount, and lets them potentially choose another payment method. Also, now it is translatable.

#### Strategy for handling the error and displaying fields

Stripe provides a "hard-coded" list of minimum amounts per currency, however it depends on the currencies of the merchant's bank accounts (deposit currencies). For example:

1. A customer browses the store in EUR. There's a minimum amount of EUR 0.50 for transactions.
2. The merchant does not have an EUR bank account, but has an USD default account.
3. The minimum amount of the purchase is actually $0.50, because of the merchant's account, and EUR 0.50 would have to be converted to USD in order to determine the minimum account.

In order to strictly reject orders with less than the minimum amount, this would ideally require:

1. All minimum amounts per currency to be hard-coded **and maintained** on the server.
2. Those amounts to be converted to all presentment currencies and delivered through an endpoint.
3. The client to query/keep those amounts regularly in order to use them during checkout.

This solution seemed tedious, so I opted to utilize the errors, returned from the Stripe API, and cache their values in order to minimize the number of hoops during checkout.

### Changes proposed in this Pull Request

Handle `amount_too_small` errors, returned from the server:

- In `WC_Payment_Gateway_WCPay::process_payment` in order to:
    1. **Format** the error message correctly
    2. **Cache** the required minimum account for the given currency
- In `UPE_Payment_Gateway::create_payment_intent` in order to:
    1. **Cache** the required minimum account for the given currency
    2. **Temporarily increase** the amount for the intent in order to match the minimum amount. If the minimum is not reached when trying to update the intent later, there will be a separate error.
- Even though theoretically this point should not be reached because of the checks in  `create_payment_intent` (2. above), `UPE_Payment_Gateway::process_payment` in order to:
    1. **Format** the error message correctly
    2. **Cache** the required minimum account for the given currency

The cached value is used the following places in order to display a message without having to reach the Stripe API first:
- `WC_Payment_Gateway_WCPay::process_payment_for_order`
- `UPE_Payment_Gateway::create_payment_intent`
- `UPE_Payment_Gateway::process_payment` in case shipping costs have not increased the order total past the minimum required amount.

### Testing instructions

Preparations:

- Create or modify a physical product in order to have it priced below the threshold. Example: 0.40 EUR or USD
- Create a shipping method (ex. flax rate) priced low enough to not pass the threshold in combination with the product above. Example: 0.05 EUR or USD
- Create another shipping method, priced high enough in order to pass the threshold: 0.5 or more USD or EUR.

#### Standard flow

With UPE fields disabled, use classic checkout to:

1. Add the cheap product to the cart.
2. Navigate to the checkout page.
3. Choose the cheap shipping method and try to check out.
4. Ensure that the error shown is descriptive about the issue (see screenshots above).
5. Select the expensive shipping method.
6. Make sure the order goes through.

Now **enable UPE fields** and:

1. Add the cheap product to the cart.
2. Navigate to the checkout page.
3. **Make sure that UPE fields are visible**, meaning that the minimum amount was handled and temporarily increased.
4. Select the cheap shipping method is selected, and try to check out.
5. Verify that a proper error has been displayed.
6. Select the expensive shipping method and try checking out again.
7. Make sure that checkout goes though.

This should work with both the classic checkout and the checkout block. However, the checkout block will only display the correct message once https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4700 has been merged.

-------------------

- [x] Added changelog entry
- [x] Covered with tests
- [x] Tested on mobile does not apply

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

RadoslavGeorgiev committed:
Initial handling of <50 cents

- Handle exceptions from the API about minimum amounts and cache those amounts.
- Increase the initial intent amount in order to allow checkout to display payment (UPE) fields.
- Manually trigger errors upon payment if the amount is cached.

ToDo:

- Handle exceptions upon intent updates (just before purchase).

RadoslavGeorgiev committed:
Handling exceptions when payment intents get updated.

RadoslavGeorgiev committed:
Making extract_minimum_amount work for various currencies.

RadoslavGeorgiev committed:
Adding all possible scenarios for extract_minimum_amount as tests.

RadoslavGeorgiev committed:
Merge branch 'develop' into fix/2700-50-cents

RadoslavGeorgiev committed:
Adding the remaining tests.

RadoslavGeorgiev committed:
Simply returning false when failed transactions limits is reached.

`process_payment` can return an array, contaning `success` === true, everything else is actually ignored.
This way the syntax is shorter and conforms to the rest of the files.

RadoslavGeorgiev committed:
Adding minimum amount handling to the main gateway.

Moved the basic functions from the UPE gateway to the main one, and moved tests accordingly.

RadoslavGeorgiev committed:
Adding tests for the main gateway.

RadoslavGeorgiev committed:
Fixing tests.

The new tests were broken becaouse of left-overs from previous tests.

RadoslavGeorgiev committed:
Adding changelog entry.

RadoslavGeorgiev committed:
Reworking returned values and messages to work with the checkout block.

The checkout block displays notices differently, which is why some tags from `wc_price` needed to be stripped,
and `false` is now replaced by `[result => fail, redirect => '']`.

naman03malhotra commented on tests/unit/test-class-wc-payment-gateway-wcpay.php:
```diff
@@ -1473,4 +1477,155 @@ public function test_payment_gateway_disabled_for_unsupported_currency() {
+	public function currencies_for_minimum_accounts_provider() {
+		return [
+			[ 'usd', '$0.50', 50 ],
+			[ 'aed', '2.00 د.إ', 200 ],
```
looks like a typo in aed currency.

RadoslavGeorgiev replied:
It's not a typo 😅 

This is caused by the RTL direction of the text, and I though there is a typo too. Then I saw that editors (VS Code, Atom, and TextEdit) are having issues formatting code like this. Turns out, there is no typo, and this is properly working code, even though it looks like it isn't. Not sure why, but it is fully functional, and works as expected.

naman03malhotra replied:
TIL, will read more about it.

naman03malhotra reviewed (commented):
The proposed solution smartly handles the use-case where you are leveraging stipe's error message to extract the minimum required amount. Prevents overengineering of things, I really liked it.

The only caveat I see is the different formats of messages returned by stripe, the current regex handles only a particular type of message format.

For example, I had set up my bank account in the US, and the store currency was EUR, and for some reason, it was throwing an error. Using the debugger I identified the error message we received from the stripe as _"Error: Amount must convert to at least 50 cents. €0.40 converts to approximately $0.47."_ Which leads to the following error.

![image](https://user-images.githubusercontent.com/15019298/132869981-b63de56e-cfde-4ef2-8f50-232d8add702d.png)

Of course, an API for the same would be much more reliable than extracting values from a string, but not sure stripe provides one for this use case.

The standard message which ends with the currency code seems to work well. But I am not sure if this is the only format in which stripe returns an error.
![image](https://user-images.githubusercontent.com/15019298/132870749-3903d224-2f66-4c68-a6ad-89881e3641f8.png)

Top of my head a potential solution would be to:
- add a fallback for the hardcoded values in case the regex fails to match.
- Identify all the error message formats in this case (not sure how) and add relevant regex to cover all the cases.

RadoslavGeorgiev committed:
Making sure that unhandled minimum_amount exceptions are not forgotten.

RadoslavGeorgiev committed:
Consolidating the failed response format into a new method to avoid repetition.

RadoslavGeorgiev commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -978,10 +983,24 @@ public function process_payment_for_order( $cart, $payment_information, $additio
+					'result'   => 'fail',
+					'redirect' => '',
+				];
+			}
```
Done in https://github.com/Automattic/woocommerce-payments/pull/2872/commits/63e3a6a38e76ab35ba0927a8bf5263d30a75f85c

RadoslavGeorgiev commented on includes/class-logger.php:
```diff
@@ -68,6 +68,10 @@ public static function init_logger() {
 	 * @return bool Depending on the enable_logging setting.
 	 */
 	public static function can_log() {
+		if ( apply_filters( 'wcpay_force_logging', false ) ) {
```
It was done in order to allow the logger to actually log. Otherwise, during testing, the logging environment is not set up and it's impossible to test whether something was logged.

This could have been done earlier, but it seems like none of the other tests necessitated it.

RadoslavGeorgiev commented:
So @naman03malhotra, I checked the server, just to be sure, and it there are _definitely_ no additional details in the response. This means that in order to really solve the issue, we will need to implement additional logic within the server, which keeps both the minimum amounts, and conversion rates in mind.

We should do that, but in the meantime, the contents of this PR could help some merchants, whose customers shop in the same currency as the merchant's deposit currency (or currencies).

What do you think?

naman03malhotra commented:
Thanks for the detailed response Rado!

**What about this approach:**
What if instead of using the minimum amount, we create the payment intent with a common higher amount that works for all the currencies (which we currently support), just so that fields are visible when the intent is created.

Later while processing the payment, the payment intent will be updated with the actual amount (which includes the shipping cost as well), if the calculated amount is lower than the minimum amount, Stripe will throw an error as usual.

Just a thought, do you see any caveat in this approach?

---

This time Stripe is throwing a different error. 😓 
Which goes away after adding an item with a higher amount.



![image](https://user-images.githubusercontent.com/15019298/133330115-03de17c5-5b3e-4c69-89d2-60cd26da90b1.png)


![image](https://user-images.githubusercontent.com/15019298/133329342-3e77c6d7-fe82-4ba1-8e4f-cf864fb7b94f.png)

RadoslavGeorgiev committed:
Removing the amount extraction regex in favor of exception data.

RadoslavGeorgiev committed:
Removing leftovers from logging.

There was a specific log for whenever the message could not be parsed, but that functionality is gone now.

RadoslavGeorgiev committed:
Adding a new `Amount_Too_Small_Exception` class for better readability.

RadoslavGeorgiev committed:
Switching from notices to exceptions for better handling in the checkout block.

RadoslavGeorgiev committed:
Updated tests

RadoslavGeorgiev committed:
Moving the currency of the minimum amount to the exception

RadoslavGeorgiev committed:
Merge branch 'develop' into fix/2700-50-cents

RadoslavGeorgiev committed:
Temporarily adding a custom server branch.

RadoslavGeorgiev committed:
Merge branch 'develop' into fix/2700-50-cents

RadoslavGeorgiev committed:
Merge branch 'develop' into fix/2700-50-cents

RadoslavGeorgiev committed:
Merge trunk v3.2.0 into develop

RadoslavGeorgiev committed:
Merge branch 'develop' into fix/2700-50-cents

RadoslavGeorgiev committed:
Merge branch 'develop' into fix/2700-50-cents

RadoslavGeorgiev commented:
Closing in favor of https://github.com/Automattic/woocommerce-payments/pull/3319
